### PR TITLE
Problem: parsing commandline is hard

### PR DESCRIPTION
**Pull request by vyskocilm**
Solution: add ligthweight and minimal API for accessing argv in the most
ergonomic, yet very flexible, way. Class zargs_t distinguish between
(positional) arguments and (named) parameters. And provide different
czmq compatible API for accessing those.

It aims to be getopt compatible, so comes with -- as a way to
distinguish between arguments and parameters. In future it will handle
the --with=value and -Wvalue variants as well.

Note: this is NOT a replacement for declarative way of defining command line api (like getopt, argparse, or doctopt). This is simpler, platform independent, CLASS/czmq comaptible API for accessing your argv.

# Pull Request Notice

Before sending a pull request make sure each commit solves one clear, minimal,
plausible problem. Further each commit should have the following format:

```
Problem: X is broken

Solution: do Y and Z to fix X
```

Please avoid sending a pull request with recursive merge nodes, as they
are impossible to fix once merged. Please rebase your branch on
zeromq/czmq master instead of merging it.

git remote add upstream git@github.com:zeromq/czmq.git
git fetch upstream
git rebase upstream/master
git push -f

In case you already merged instead of rebasing you can drop the merge commit.

git rebase -i HEAD~10

Now, find your merge commit and mark it as drop and save. Finally rebase!

If you are a new contributor please have a look at our contributing guidelines:
[CONTRIBUTING.md](https://github.com/zeromq/czmq/blob/master/CONTRIBUTING.md)

### CODE REVIEW

**Comment by vyskocilm** [comment]
cc @sappo  @bluca please review the API, however I hope it makes a sense!

File: project.xml [hunk 1]
```diff
@@ -13,10 +13,10 @@
     <use project = "libzmq" />
     <use project = "uuid" optional = "1" implied = "1" />
     <use project = "systemd" optional = "1" implied = "1" min_major = "200" />
-    <target name = "*" />
```
Comment (bluca) [inline project.xml]: any particular reason to remove this?

File: builds/gyp/project.gyp [hunk 2]
```diff
@@ -138,7 +140,6 @@
         '../../src/czmq_classes.h'
       ],
       'dependencies': [
-        '../../../libzmq/builds/gyp/project.gyp:libzmq',
```
Comment (bluca) [inline builds/gyp/project.gyp]: This was probably done by zproject, don't think it's right to remove it?

**Comment by bluca** [comment]
Looks great, just noticed two minor things that were probably caused by zproject

File: project.xml (hunk 1 above)
Comment (vyskocilm) [inline project.xml]: Yes, my zproject rpm was broken ... I did drop it in order to continue with a development. Will bring it back.

**Comment by vyskocilm** [comment]
@bluca thanks for spotting it - I put it back and did gsl project.xml. As I am using zproject from OBS, I believe this is way to go on Linux

**Comment by bluca** [comment]
Great, thanks! The API looks fine to me.

There seems to be a problem with Python bindings though: https://travis-ci.org/zeromq/czmq/jobs/239377425

**Comment by bluca** [comment]
Seems strange, not sure why Python is complaining about zargs_lookup, doesn't seem to be anything out of the ordinary

**Comment by bluca** [comment]
Ah I see the problem, 2 APIs have the name set to zargs_lookup[x] in the API: https://github.com/zeromq/czmq/blob/master/api/zargs.api#L79
So that's how they get generated and imported and exported, but the actual implementation calls them zargs_param_lookup[x]: https://github.com/zeromq/czmq/blob/master/src/zargs.c#L187

So the zargs_lookup[x] symbols do not exist in the shared library since there's no function with those names, and thus Python fails to import them

Could you please fix it?

**Comment by vyskocilm** [comment]
https://github.com/zeromq/czmq/pull/1680

**Pull request merged**